pkg/gofr/datasource/file/ftp/fs.go

```diff
@@ -0,0 +1,353 @@
+package ftp
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"net/textproto"
+	"os"
+	"path"
+	"slices"
+	"time"
+
+	"gofr.dev/pkg/gofr/container"
+
+	"github.com/jlaffaye/ftp"
+)
+
+// Conn struct embeds the *ftp.ServerConn returned by ftp server on successful connection.
+type Conn struct {
+	*ftp.ServerConn
+}
+
+// Retr wraps the ftp retrieve method to return a ftpResponse interface type.
+func (c *Conn) Retr(filepath string) (ftpResponse, error) {
+	return c.ServerConn.Retr(filepath)
+}
+
+func (c *Conn) RetrFrom(filepath string, offset uint64) (ftpResponse, error) {
+	return c.ServerConn.RetrFrom(filepath, offset)
+}
+
+var (
+	errEmptyFilename  = errors.New("filename cannot be empty")
+	errEmptyPath      = errors.New("file/directory path cannot be empty")
+	errEmptyDirectory = errors.New("directory name cannot be empty")
+	errInvalidArg     = errors.New("invalid filename/directory")
+)
+
+// fileSystem represents a file system interface over FTP.
+type fileSystem struct {
+	*file
+	conn    ServerConn
+	config  *Config
+	logger  Logger
+	metrics Metrics
+}
+
+// Config represents the FTP configuration.
+type Config struct {
+	Host      string // FTP server hostname
+	User      string // FTP username
+	Password  string // FTP password
+	Port      int    // FTP port
+	RemoteDir string // Remote directory path. Base Path for all FTP Operations.
+}
+
+// New initializes a new instance of ftpFileSystem with provided configuration.
+func New(config *Config) container.FileSystemProvider {
+	return &fileSystem{config: config}
+}
+
+// UseLogger sets the Logger interface for the FTP file system.
+func (f *fileSystem) UseLogger(logger interface{}) {
+	if l, ok := logger.(Logger); ok {
+		f.logger = l
+	}
+}
+
+// UseMetrics sets the Metrics interface.
+func (f *fileSystem) UseMetrics(metrics interface{}) {
+	if m, ok := metrics.(Metrics); ok {
+		f.metrics = m
+	}
+}
+
+// Connect establishes a connection to the FTP server and logs in.
+func (f *fileSystem) Connect() {
+	ftpServer := fmt.Sprintf("%v:%v", f.config.Host, f.config.Port)
+
+	defer f.processLog(&FileLog{Operation: "Connect", Location: ftpServer}, time.Now())
+
+	const dialTimeout = 5 * time.Second
```

Comment:
I'm still surprised this to be hard coded here, and not available as a configuration parameter, even if the default could be 5s

Reply:
Ok, will be moving it to the configs.